enhancements/sig-policy/122-expanded-hub-templates-access/README.md

```diff
@@ -0,0 +1,310 @@
+that I can stop defining additional policies on `local-cluster` to sync the required information to the root policy
+namespace.
+
+### Implementation Details/Notes/Constraints
+
+Although the user experience is rather simple, there's complexity in the implementation. One aspect is that the number
+of service accounts is not known at startup and can constantly change. It is also a bit complex to prevent duplication
+of Kubernetes API watches, underlying TCP connections, and tokens if multiple policies reference the same service
+account.
+
+Currently, the Governance Policy Propagator only uses a single `TemplateResolver` client from the
+[go-template-utils](https://github.com/stolostron/go-template-utils) library and sets restrictions based on the root
+policy namespace. That will still remain the default, however, we'll need exactly one `TemplateResolver` client per
+service account. To do this, I recommend having a new `sync.Map` field named `TemplateResolvers` on the
+`ReplicatedPolicyReconciler` struct and each key would be a `NamespacedName` from `k8s.io/apimachinery/pkg/types`
+representing the service account. Each value would be an instance of the following struct:
+
+```go
+type templateResolverWithCancel struct {
+	lock           *sync.RWMutex
+	cancel         context.CancelFunc
+	resolver       *templates.TemplateResolver
+	dynamicWatcher k8sdepwatches.DynamicWatcher
+}
+```
+
+When the replicated policy controller with the Governance Policy Propagator encounters a
+`spec.hubTemplateOptions.serviceAccountName` field, it'll perform an operation similar to this to either create the
+`TemplateResolver` if it's the first reconcile for that `serviceAccountName` or to use the preexisting one:
+
+```go
+var templateResolver *templates.TemplateResolver
+
+nsName := types.NamespacedName{
+	Namespace: rootPlc.Namespace, Name: replicatedPlc.Spec.HubTemplateOptions.ServiceAccountName,
+}
+
+resolver := &templateResolverWithCancel{
+	lock: &sync.RWMutex{},
+}
+
+resolver.lock.Lock()
+
+oldResolver, loaded := r.TemplateResolvers.LoadOrStore(nsName, resolver)
+if loaded {
+	resolver.lock.Unlock()
+
+	oldResolverTyped := oldResolver.(*templateResolverWithCancel)
+	oldResolverTyped.lock.RLock()
+
+	// Handle the case where resolver is getting cleaned up while the lock is held.
+	// Just try again in this case.
+	if oldResolverTyped.resolver == nil {
+		oldResolverTyped.lock.RUnlock()
+
+		// TODO: Write code to handle repeating the above code block
+	}
+
+	defer oldResolverTyped.lock.RUnlock()
+
+	templateResolver = oldResolverTyped.resolver
+
+	break
+}
+
+// TODO: Code to get a token and instantiate a TemplateResolver instance for the service account and
+// set the values on `resolver`.
+```
+
+There should be a goroutine that periodically cleans up entries in `TemplateResolvers` that are for service accounts no
+longer referenced by any root policy. This is simpler than managing a reference count for each service account.
```

Comment:
Maybe if there will be a goroutine to clean things up anyway, there should just be that goroutine fully managing all of the entries, and we use channels to synchronize work. Just a thought - I'm not sure exactly what it would look like.

Reply:
It's an interesting idea but I think it'd be more complex than what is suggested in the design since you'd need a channel to request the template resolver for the SA and then a channel to receive it. Each reconcile call would somehow need to pick the correct channel to listen on to get the correct template resolver. I think a `sync.Map` and a `sync.RWMutex` to synchronize the creation but read concurrently is easier to understand and likely more efficient.

Let me know if you had something else in mind than what I pictured!